MvvmCross/Core/Core/ViewModels/MvxViewModel.cs

```diff
@@ -82,10 +83,56 @@ public abstract class MvxViewModel<TInit> : MvxViewModel, IMvxViewModelInitializ
                 throw new MvxIoCResolveException("There is no implementation of IMvxJsonConverter registered. You need to use the MvvmCross Json plugin or create your own implementation of IMvxJsonConverter.");
             }
 
-            var deserialized = serializer.DeserializeObject<TInit>(parameter);
-            await Init(deserialized);
+            var deserialized = serializer.DeserializeObject<TParameter>(parameter);
+            await Initialize(deserialized);
+        }
+
+        public abstract Task Initialize(TParameter parameter);
+    }
+
+    //TODO: Not possible to name MvxViewModel, name is MvxViewModelResult for now
+    public abstract class MvxViewModelResult<TResult> : MvxViewModel, IMvxViewModelResult<TResult> where TResult : class
+    {
+        TaskCompletionSource<TResult> _tcs;
+
+        public void SetClose(TaskCompletionSource<TResult> tcs)
+        {
+            _tcs = tcs;
+        }
+
+        public virtual Task<bool> Close(TResult result)
+        {
+            _tcs.TrySetResult(result);
+            return Task.FromResult(Close(this));
+        }
+
+		public override void Disappeared()
+		{
+			_tcs?.TrySetCanceled();
+			base.Disappeared();
+		}
+    }
+
+    public abstract class MvxViewModel<TParameter, TResult> : MvxViewModel, IMvxViewModel<TParameter, TResult> where TParameter : class where TResult : class
+    {
+        TaskCompletionSource<TResult> _tcs;
+
+        public void SetClose(TaskCompletionSource<TResult> tcs)
+        {
+            _tcs = tcs;
         }
 
-        public abstract Task Init(TInit parameter);
+        public abstract Task Initialize(TParameter parameter);
+        public virtual Task<bool> Close(TResult result)
+        {
+            _tcs.TrySetResult(result);
```

Comment:
Below you null check `_tcs` but that is not done here.  I think we need to decide on the contract and be consistent.